File: src/ui/windows/Sidebar.cs
using System;
using System.Windows.Controls;
using System.Windows.Media;
using ExcelCore;
using System.Collections.Generic;

namespace create_excel_v2_dru9yl.src.ui.windows
{
    /// <summary>
    /// Represents the Sidebar in Excel, providing access to various panes and tools
    /// </summary>
    public class Sidebar : UserControl
    {
        private StackPanel sidebarPanel;
        private ExcelCore.Workbook currentWorkbook;
        private Dictionary<string, UserControl> sidebarPanes;

        /// <summary>
        /// Initializes a new instance of the Sidebar class
        /// </summary>
        /// <param name="workbook">The current ExcelCore.Workbook</param>
        public Sidebar(ExcelCore.Workbook workbook)
        {
            // Initialize the sidebar panel
            InitializeSidebar();

            // Set the current workbook
            currentWorkbook = workbook;

            // Create and add sidebar panes
            CreateSidebarPanes();

            // Set up event handlers for sidebar interactions
            SetupEventHandlers();
        }

        /// <summary>
        /// Initializes the Sidebar control and its components
        /// </summary>
        private void InitializeSidebar()
        {
            // Create a new StackPanel for the sidebar
            sidebarPanel = new StackPanel();

            // Set sidebar properties (e.g., width, background color)
            sidebarPanel.Width = 250;
            sidebarPanel.Background = Brushes.WhiteSmoke;

            // Add the sidebar panel to the UserControl
            this.Content = sidebarPanel;

            // Apply styling to match Excel's theme
            ApplyExcelTheme();
        }

        /// <summary>
        /// Creates and initializes the various sidebar panes
        /// </summary>
        private void CreateSidebarPanes()
        {
            sidebarPanes = new Dictionary<string, UserControl>();

            // Create instances of different sidebar panes (e.g., FormatPane, ReviewPane)
            // Add created panes to the sidebarPanes dictionary
            // Create toggle buttons for each pane

            // Example:
            // sidebarPanes.Add("Format", new FormatPane());
            // sidebarPanes.Add("Review", new ReviewPane());

            foreach (var pane in sidebarPanes)
            {
                Button toggleButton = new Button();
                toggleButton.Content = pane.Key;
                toggleButton.Click += (sender, e) => ToggleSidebarPane(pane.Key);
                sidebarPanel.Children.Add(toggleButton);
            }
        }

        /// <summary>
        /// Toggles the visibility of a specific sidebar pane
        /// </summary>
        /// <param name="paneName">The name of the pane to toggle</param>
        private void ToggleSidebarPane(string paneName)
        {
            // Check if the requested pane exists
            if (sidebarPanes.TryGetValue(paneName, out UserControl pane))
            {
                // If visible, hide the current pane
                if (pane.Visibility == System.Windows.Visibility.Visible)
                {
                    pane.Visibility = System.Windows.Visibility.Collapsed;
                }
                else
                {
                    // If hidden, show the requested pane and hide others
                    foreach (var otherPane in sidebarPanes.Values)
                    {
                        otherPane.Visibility = System.Windows.Visibility.Collapsed;
                    }
                    pane.Visibility = System.Windows.Visibility.Visible;
                }

                // Update toggle button states
                UpdateToggleButtonStates();
            }
        }

        /// <summary>
        /// Updates the sidebar content based on the current workbook state
        /// </summary>
        public void UpdateSidebar()
        {
            // Get the active worksheet and selected cells
            ExcelCore.Worksheet activeWorksheet = currentWorkbook.ActiveWorksheet;
            IEnumerable<ExcelCore.Cell> selectedCells = activeWorksheet.GetSelectedCells();

            // Update each sidebar pane with current context
            foreach (var pane in sidebarPanes.Values)
            {
                // Assuming each pane has an UpdateContent method
                // (pane as IUpdatable)?.UpdateContent(activeWorksheet, selectedCells);
            }

            // Refresh the visibility of panes based on current context
            RefreshPaneVisibility(activeWorksheet, selectedCells);
        }

        private void ApplyExcelTheme()
        {
            // Apply Excel-like styling to the sidebar
        }

        private void SetupEventHandlers()
        {
            // Set up event handlers for sidebar interactions
        }

        private void UpdateToggleButtonStates()
        {
            // Update the state of toggle buttons based on pane visibility
        }

        private void RefreshPaneVisibility(ExcelCore.Worksheet activeWorksheet, IEnumerable<ExcelCore.Cell> selectedCells)
        {
            // Implement logic to show/hide panes based on current context
        }

        // Human tasks:
        // TODO: Implement a customization feature for users to add/remove sidebar panes
        // TODO: Add support for resizing the sidebar and individual panes
        // TODO: Implement drag-and-drop functionality for rearranging sidebar panes
        // TODO: Add support for collapsible sections within sidebar panes
        // TODO: Implement a search functionality within the sidebar
        // TODO: Add support for pinning frequently used panes
        // TODO: Implement context-sensitive help for sidebar pane features
        // TODO: Add support for custom user-created panes
        // TODO: Implement accessibility features for screen readers
        // TODO: Add support for theming and custom color schemes in the sidebar
        // TODO: Ensure proper scaling of sidebar for different DPI settings
        // TODO: Implement smooth animations for sidebar show/hide transitions
        // TODO: Design and implement additional sidebar panes for Excel functionality
        // TODO: Ensure consistent styling across all sidebar panes
        // TODO: Implement smooth transitions when switching between panes
        // TODO: Add keyboard shortcuts for toggling sidebar panes
        // TODO: Implement context-sensitive pane visibility logic
        // TODO: Optimize update performance for large selections
    }
}

File: src/ui/macos/Sidebar.swift
import AppKit
import ExcelCore

class Sidebar: NSView {
    private var stackView: NSStackView
    private var currentWorkbook: ExcelCore.Workbook
    private var sidebarPanes: [String: NSView]
    
    init(workbook: ExcelCore.Workbook) {
        // Initialize properties
        self.currentWorkbook = workbook
        self.stackView = NSStackView()
        self.sidebarPanes = [:]
        
        // Call superclass initializer
        super.init(frame: NSRect.zero)
        
        // Set up the UI
        setupUI()
        
        // Create and add sidebar panes
        createSidebarPanes()
        
        // Set up the layout of the sidebar
        self.addSubview(stackView)
        stackView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            stackView.topAnchor.constraint(equalTo: self.topAnchor),
            stackView.leadingAnchor.constraint(equalTo: self.leadingAnchor),
            stackView.trailingAnchor.constraint(equalTo: self.trailingAnchor),
            stackView.bottomAnchor.constraint(equalTo: self.bottomAnchor)
        ])
        
        // Configure event handlers
        NotificationCenter.default.addObserver(self, selector: #selector(updateSidebar), name: NSNotification.Name("WorkbookStateChanged"), object: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupUI() {
        // Create and configure the NSStackView
        stackView = NSStackView(frame: self.bounds)
        stackView.orientation = .vertical
        stackView.spacing = 8
        stackView.distribution = .fillEqually
        
        // Apply styling to match Excel's theme
        self.wantsLayer = true
        self.layer?.backgroundColor = NSColor(calibratedWhite: 0.95, alpha: 1.0).cgColor
    }
    
    private func createSidebarPanes() {
        // Create instances of different sidebar panes
        let formatPane = FormatPane()
        let reviewPane = ReviewPane()
        // Add more panes as needed
        
        // Add created panes to the sidebarPanes dictionary
        sidebarPanes["Format"] = formatPane
        sidebarPanes["Review"] = reviewPane
        
        // Create toggle buttons for each pane
        for (paneName, pane) in sidebarPanes {
            let toggleButton = NSButton(title: paneName, target: self, action: #selector(paneToggleClicked(_:)))
            toggleButton.setButtonType(.toggle)
            stackView.addArrangedSubview(toggleButton)
        }
    }
    
    @objc private func paneToggleClicked(_ sender: NSButton) {
        toggleSidebarPane(paneName: sender.title)
    }
    
    private func toggleSidebarPane(paneName: String) {
        guard let pane = sidebarPanes[paneName] else { return }
        
        // If visible, hide the current pane
        if pane.isHidden == false {
            pane.isHidden = true
            return
        }
        
        // If hidden, show the requested pane and hide others
        for (name, view) in sidebarPanes {
            view.isHidden = (name != paneName)
        }
        
        // Update toggle button states
        for subview in stackView.arrangedSubviews {
            if let button = subview as? NSButton {
                button.state = button.title == paneName ? .on : .off
            }
        }
    }
    
    @objc private func updateSidebar() {
        guard let activeWorksheet = currentWorkbook.activeWorksheet,
              let selectedRange = activeWorksheet.selectedRange else { return }
        
        // Update each sidebar pane with current context
        for (_, pane) in sidebarPanes {
            if let updatable = pane as? SidebarPaneUpdatable {
                updatable.update(worksheet: activeWorksheet, selectedRange: selectedRange)
            }
        }
        
        // Refresh the visibility of panes based on current context
        // This is a placeholder for context-sensitive pane visibility logic
        // Implement your own logic here based on the current workbook state
    }
}

// Protocol for updatable sidebar panes
protocol SidebarPaneUpdatable {
    func update(worksheet: ExcelCore.Worksheet, selectedRange: ExcelCore.Range)
}

// MARK: - Human Tasks
/*
 TODO: Implement a customization feature for users to add/remove sidebar panes
 TODO: Add support for resizing the sidebar and individual panes
 TODO: Implement drag-and-drop functionality for rearranging sidebar panes
 TODO: Add support for collapsible sections within sidebar panes
 TODO: Implement a search functionality within the sidebar
 TODO: Add support for pinning frequently used panes
 TODO: Implement context-sensitive help for sidebar pane features
 TODO: Add support for custom user-created panes
 TODO: Implement accessibility features for VoiceOver
 TODO: Add support for Touch Bar integration for sidebar functionality
 */